### PR TITLE
Hamming distance between two strings (Compares Unicode code points)

### LINKED ISSUE #836: [RFC]: Hamming distance between two strings
### Description

This RFC proposes adding a function to calculate the Hamming distance between two strings.

The function should have the following signature `(a: string, b: string): number`.

The function should take two strings as arguments and return the Hamming distance between them. The Hamming distance is defined as the number of characters that have to be changed to convert one string to the other. Since it only allows substitutions, it can only be used to compare strings of the same length.

Additionally, in order to account for code points and grapheme clusters, we should add separate packages for dealing with each, as the underlying algorithms are likely to differ. We then can provide a more general API which unifies the underlying algorithms. Accordingly, we should create the following packages:

- [x] `@stdlib/string/base/distances/hamming`: compares UTF-16 code units.
    - PR: https://github.com/stdlib-js/stdlib/pull/1166
- [ ] `@stdlib/string/base/distances/hamming-code-points`: compares Unicode code points.
- [ ] `@stdlib/string/base/distances/hamming-grapheme-clusters`: compares grapheme clusters (i.e., visual characters)

Once the above are completed, we can add

- [ ] `@stdlib/string/distances/hamming`: unifies the above "base" packages and provides an option for specifying the computation "mode" (i.e., `code_units`, `code_points`, or `grapheme_clusters`, with `grapheme_clusters` being the default).

### Related Issues

Related issues https://github.com/stdlib-js/stdlib/issues/151.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by JayPokale**
## Description
Calculates the Hamming Code Points Distance distance between two equal-length unicode code points.

> Does this pull request have any related issues?

This pull request:

-   progresses #836

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by Pranavchiku** [comment]
@JayPokale can you please ensure linters are set correctly in your device. I think those are not.

File: CONTRIBUTORS [hunk 1]
```diff
@@ -24,6 +24,7 @@ Golden <103646877+AuenKr@users.noreply.github.com>
 Harshita Kalani <harshitakalani02@gmail.com>
 Jaimin Godhani <112328542+Jai0401@users.noreply.github.com>
 James Gelok <jdgelok@gmail.com>
+Jay Pokale <jay.pokale.35@gmail.com>
```
Comment (Pranavchiku) [inline CONTRIBUTORS]: This is automated, please revert this commit.

File: lib/node_modules/@stdlib/string/base/distances/hamming-code-points/benchmark/benchmark.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      b.fail("should return a number");
+    }
+  }
+  b.toc();
+  if (typeof out !== "number") {
+    b.fail("should return a number");
+  }
+  b.pass("benchmark finished");
+  b.end();
+});
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/base/distances/hamming-code-points/benchmark/benchmark.js:62]: This file should be TAB indented.

File: lib/node_modules/@stdlib/string/base/distances/hamming-code-points/docs/types/test.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ * limitations under the License.
+ */
+
+import hammingCodePointsDistance from "./index";
+
+// TESTS //
+
+// The function returns a number...
+{
+  hammingCodePointsDistance("", ""); // $ExpectType number
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/base/distances/hamming-code-points/docs/types/test.ts]:
```suggestion
  hammingCodePointsDistance( "", "" ); // $ExpectType number
```

File: lib/node_modules/@stdlib/string/base/distances/hamming-code-points/docs/types/test.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+// The compiler throws an error if the function is provided a first argument which is not a string...
+{
+  hammingCodePointsDistance(true, ""); // $ExpectError
+  hammingCodePointsDistance(false, ""); // $ExpectError
+  hammingCodePointsDistance(null, ""); // $ExpectError
+  hammingCodePointsDistance(undefined, ""); // $ExpectError
+  hammingCodePointsDistance(5, ""); // $ExpectError
+  hammingCodePointsDistance([], ""); // $ExpectError
+  hammingCodePointsDistance({}, ""); // $ExpectError
+  hammingCodePointsDistance((x: number): number => x, ""); // $ExpectError
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/base/distances/hamming-code-points/docs/types/test.ts]: same comment here

File: lib/node_modules/@stdlib/string/base/distances/hamming-code-points/docs/types/test.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+// The compiler throws an error if the function is provided a second argument which is not a string...
+{
+  hammingCodePointsDistance("", true); // $ExpectError
+  hammingCodePointsDistance("", false); // $ExpectError
+  hammingCodePointsDistance("", null); // $ExpectError
+  hammingCodePointsDistance("", undefined); // $ExpectError
+  hammingCodePointsDistance("", 5); // $ExpectError
+  hammingCodePointsDistance("", []); // $ExpectError
+  hammingCodePointsDistance("", {}); // $ExpectError
+  hammingCodePointsDistance("", (x: number): number => x); // $ExpectError
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/base/distances/hamming-code-points/docs/types/test.ts]: here as well

File: lib/node_modules/@stdlib/string/base/distances/hamming-code-points/docs/types/test.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+// The compiler throws an error if the function is provided an unsupported number of arguments...
+{
+  hammingCodePointsDistance(); // $ExpectError
+  hammingCodePointsDistance(""); // $ExpectError
+  hammingCodePointsDistance("", "", 3); // $ExpectError
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/base/distances/hamming-code-points/docs/types/test.ts]: sme comment

File: lib/node_modules/@stdlib/string/base/distances/hamming-code-points/docs/types/test.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
+/*
+ * @license Apache-2.0
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/base/distances/hamming-code-points/docs/types/test.ts]: this file shall be TAB indented

File: lib/node_modules/@stdlib/string/base/distances/hamming-code-points/lib/index.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ * // returns 3
+ *
+ * dist = hammingCodePointsDistance( "a😊c", "a😃c" );
+ * // returns 1
+ *
+ * dist = hammingCodePointsDistance( "こんにちは", "さようなら" );
+ * // returns 5
+ */
+
+// MODULES //
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/base/distances/hamming-code-points/lib/index.js]:
```suggestion
 */


// MODULES //
```

File: lib/node_modules/@stdlib/string/base/distances/hamming-code-points/lib/index.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+var main = require("./main.js");
+
+// EXPORTS //
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/base/distances/hamming-code-points/lib/index.js]:
```suggestion
var main = require("./main.js");


// EXPORTS //
```

File: lib/node_modules/@stdlib/string/base/distances/hamming-code-points/lib/main.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+ */
+
+"use strict";
+
+// MODULES //
+
+var isString = require("@stdlib/assert/is-string").isPrimitive;
+var format = require("@stdlib/string/format");
+
+// MAIN //
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/base/distances/hamming-code-points/lib/main.js]:
```suggestion
var format = require("@stdlib/string/format");


// MAIN //
```

File: lib/node_modules/@stdlib/string/base/distances/hamming-code-points/lib/main.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  var res = 0;
+  for (var i = 0; i < s1.length; i++) {
+    if (s1.codePointAt(i) !== s2.codePointAt(i)) {
+      res += 1;
+    }
+  }
+  return res;
+}
+
+// EXPORTS //
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/base/distances/hamming-code-points/lib/main.js:69]: Tab indentation required

File: lib/node_modules/@stdlib/string/base/distances/hamming-code-points/lib/main.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+ * @throws {TypeError} first argument must be a string
+ * @throws {TypeError} second argument must be a string
+ * @returns {integer} Hamming code points distance
+ *
+ * @example
+ * var distance = hammingCodePointsDistance( 'algorithm', 'altruistic' );
+ * // returns -1
+ */
+function hammingCodePointsDistance(s1, s2) {
+  if (!isString(s1)) {
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/base/distances/hamming-code-points/lib/main.js]: Please declare variables at beginning of file

File: lib/node_modules/@stdlib/string/base/distances/hamming-code-points/test/test.js [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+    for (var i = 0; i < values.length; i++) {
+      t.strictEqual(
+        hammingCodePointsDistance(values[i][0], values[i][1]),
+        expected[i],
+        "returns expected value"
+      );
+    }
+    t.end();
+  }
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/base/distances/hamming-code-points/test/test.js]: Tab indentation, bracket spacing, etc.

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Please go through development guide and iterate through the PR you referred. Setup dev environment and then this PR can be reviewed.

**Comment by JayPokale** [comment]
@Pranavchiku I converted code in required format as mentioned above.

**Comment by kgryte** [comment]
This PR is currently failing CI and needs to be updated before we can review.

**Pull request closed without merging**